tests/engine-api.md

```diff
@@ -219,3 +223,62 @@ All test cases described in this section are beginning in a post-Merge world, i.
   
   </details>
 
+
+## CL client tests
+
+* [ ] `QUANTITY` field values are encoded correctly
+  <details>
+  <summary>Click for details &#9662;</summary>
+  
+  * Payload `P1` has all `QUANTITY` field values greater than `255`
+  * CL processes `BeaconBlock(P1)`
+    * All `P1` `QUANTITY` fields are big-endian
+  
+  </details>
+
+* [ ] `INVALID` *canonical* chain payload
+  <details>
+  <summary>Click for details &#9662;</summary>
+  
+  * `INV_P` is an `INVALID` payload extending *canonical* chain
+  * CL imports `BeaconBlock(INV_P)`
+    * `BeaconBlock(INV_P)` isn't available via `GET /eth/v1/beacon/headers/{block_id}`
+    * EL mock artificially returns `INVALID`
+  
+  </details>
+
+* [ ] Invalid `block_hash`
+  <details>
+  <summary>Click for details &#9662;</summary>
+  
+  * Payload `P` is responded with `INVALID_BLOCK_HASH`
+    * EL mock artificially returns this status
+  * CL discards `BeaconBlock(P)`
+  
+  </details>
+
+* [ ] `SYNCING` with *invalid* chain
+  <details>
+  <summary>Click for details &#9662;</summary>
+  
+  * `Genesis <- P1 <- P2 <- P3 <- P4`
+  * CL imports `BeaconBlock(P1) ... BeaconBlock(P4)` block by block
+    * EL mock should respond `SYNCING` on `newPayload(P2)`
+    * EL mock should respond `status: INVALID, latestValidHash: P1.blockHash` on `newPayload(P4)`
+  * CL's head must be `BeaconBlock(P1)`
+  * EL must receive `forkchoiceUpdated(P1)`
+  * `finalized`, `safe` and head blocks are as expected
```

Comment:
This is an interesting test case. From the implementation standpoint, I think the proxy can simply select any payload as `P1`, return `SYNCING` until it counts the 4th one after that, then finally return the `INVALID`.

On a side note, I think the ELs would only be capable of returning `INVALID` up to `P3`, and from `P4` onward would return `SYNCING`, according to implementation here [Engine API: thoughts on latestValidHash](https://hackmd.io/GDc0maGsQeKfP8o2C7L52w) ?

Reply:
> I think the proxy can simply select any payload as P1, return SYNCING until it counts the 4th one after that, then finally return the INVALID.

Yeah, this should work.

> according to implementation here Engine API: thoughts on latestValidHash

Actually, according to this proposal EL should cache `invalidTipHash` which means that it will remember the head of invalid chain, this is the only way EL may link a new payload to invalid chain.